File: cpp/series/kitamasa.cpp
using ll = long long;

#define K 1003
#define MOD 1000000007
 
// (0-indexed)
// a[i] = a_i, c[i] = c_i
// a_{i+k} = c_i*a_i + c_{i+1}*a_{i+1} + ... + c_{i+k-1}*a_{i+k-1}
ll c[K], a[K];
 
// C(N, *) -> C(N+1, *)
void nxt(int k, ll c0[K], ll c1[K]) {
  c1[0] = (c0[k-1] * c[0]) % MOD;
  for(int i=0; i<k-1; ++i) {
    c1[i+1] = (c0[i] + c0[k-1]*c[i+1]) % MOD;
  }
}
 
// C(N, *) -> C(2N, *)
void dbl(int k, ll c0[K], ll c1[K]) {
  ll cs[K][K];
  for(int i=0; i<k; ++i) cs[0][i] = c0[i];
  for(int i=0; i<k-1; ++i) {
    nxt(k, cs[i], cs[i+1]);
  }
  for(int i=0; i<k; ++i) {
    c1[i] = 0;
    for(int j=0; j<k; ++j) {
      c1[i] += cs[0][j] * cs[j][i];
      c1[i] %= MOD;
    }
  }
}
 
// caluculate a_m
ll solve(int m, int k) {
  ll c0[K], c1[K];
  if(m == 0) {
    return a[0];
  }
  for(int i=0; i<k; ++i) c0[i] = 0;
  c0[1] = 1;
 
  int p = 31;
  while(((m >> --p) & 1) == 0);
  while(p-- > 0) {
    dbl(k, c0, c0);
    if((m >> p) & 1) {
      nxt(k, c0, c1);
      for(int j=0; j<k; ++j) c0[j] = c1[j];
    }
  }
 
  ll res = 0;
  for(int i=0; i<k; ++i) {
    res += c0[i]*a[i];
    res %= MOD;
  }
  return res;
}

File: python/series/kitamasa.py
# (0-indexed)
# a[i] = a_i, c[i] = c_i
# a_{i+k} = c_i*a_i + c_{i+1}*a_{i+1} + ... + c_{i+k-1}*a_{i+k-1}
def solve1(m, k, C, A):
    MOD = 10**9 + 7
    C0 = [0]*k; C1 = [0]*k
    if m == 0:
        return A[0]
    C0[1] = 1

    def inc(k, C0, C1):
        C1[0] = C0[k-1] * C[0] % MOD
        for i in range(k-1):
            C1[i+1] = (C0[i] + C0[k-1]*C[i+1]) % MOD

    def dbl(k, C0, C1):
        D0 = [0]*k; D1 = [0]*k
        D0[:] = C0[:]
        for j in range(k):
            C1[j] = C0[0] * C0[j] % MOD
        for i in range(1, k):
            inc(k, D0, D1)

            for j in range(k):
                C1[j] += C0[i] * D1[j] % MOD
            D0, D1 = D1, D0
        for i in range(k):
            C1[i] %= MOD

    p = 32
    while (m >> p) & 1 == 0:
        p -= 1

    while p:
        p -= 1
        dbl(k, C0, C1)
        C0, C1 = C1, C0

        if (m >> p) & 1:
            inc(k, C0, C1)
            C0, C1 = C1, C0

    return sum(C0[i] * A[i] for i in range(k)) % MOD

# solve1の関数を展開したバージョン
def solve2(m, k, C, A):
    MOD = 10**9 + 7

    C0 = [0]*k; C1 = [0]*k
    if m == 0:
        return A[0]
    C0[1] = 1

    p = 32
    while (m >> p) & 1 == 0:
        p -= 1

    D0 = [0]*k; D1 = [0]*k
    while p:
        p -= 1

        #dbl(k, C0, C1)
        D0[:] = C0[:]
        for j in range(k):
            C1[j] = C0[0] * C0[j] % MOD
        for i in range(1, k):
            #inc(k, D0, D1)
            D1[0] = D0[k-1] * C[0] % MOD
            for j in range(k-1):
                D1[j+1] = (D0[j] + D0[k-1]*C[j+1]) % MOD

            for j in range(k):
                C1[j] += C0[i] * D1[j] % MOD
            D0, D1 = D1, D0
        for i in range(k):
            C1[i] %= MOD
        C0, C1 = C1, C0

        if (m >> p) & 1:
            #inc(k, C0, C1)
            C1[0] = C0[k-1] * C[0] % MOD
            for i in range(k-1):
                C1[i+1] = (C0[i] + C0[k-1]*C[i+1]) % MOD
            C0, C1 = C1, C0

    return sum(C0[i] * A[i] for i in range(k)) % MOD